File: Core/Bridge/io/Mouse.cs
using Godot;

namespace monoe.exe.Core.Bridge.io;

public static class Mouse
{
  public static object[] Position()
  {
    var p = SceneRoot.I.GetViewport().GetMousePosition();
    return [(double)p.X, (double)p.Y];
  }

  public static object[] Delta()
  {
    var d = Input.GetLastMouseVelocity();
    return [(double)d.X, (double)d.Y];
  }

  public static bool ButtonPressed(string button)
  {
    return button.ToLowerInvariant() switch
    {
      "left"   => Input.IsMouseButtonPressed(MouseButton.Left),
      "right"  => Input.IsMouseButtonPressed(MouseButton.Right),
      "middle" => Input.IsMouseButtonPressed(MouseButton.Middle),
      "x1"     => Input.IsMouseButtonPressed(MouseButton.Xbutton1),
      "x2"     => Input.IsMouseButtonPressed(MouseButton.Xbutton2),
      _ => false
    };
  }
}

File: monoelib/io/mouse.lua
---@diagnostic disable: return-type-mismatch, missing-return-value
local engine = require("monoelib.engine")

monoe = monoe or {}
monoe.io = monoe.io or {}

---@class monoe.io.mouse
---Provides helpful functions that'll allow you to use the mouse during your game.
monoe.io.mouse = {}

local MOUSE = "monoe.exe.Core.Bridge.io.Mouse"

---Returns the mouse position in viewport space.
---@return number x, number y
function monoe.io.mouse.position()
  return engine.staticcall(MOUSE, "Position")
end

---Returns the mouse movement delta since last frame.
---@return number dx, number dy
function monoe.io.mouse.delta()
  return engine.staticcall(MOUSE, "Delta")
end

---Returns true if a mouse button is pressed.
---@param button "left"|"right"|"middle"|"x1"|"x2"
---@return boolean
function monoe.io.mouse.down(button)
  return engine.staticcall(MOUSE, "ButtonPressed", button)
end

_G.monoe = monoe

return monoe.io.mouse